Initial PyTorch support

NikEfth opened the pull request:
<!-- Fill in most of this text, and delete what is not appropriate.
Please read and adhere to the [contribution guidelines](https://github.com/UCL/STIR/blob/master/CONTRIBUTING.md).
Did you sign the STIR Contribution License Agreement?
-->

## Changes in this pull request

* Link and compile with PyTorch
* 

## Testing performed

* make tests

## Related issues
<!-- Use keywords such as "fixes", "closes", see https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue -->


## Checklist before requesting a review
<!--Put an x between the [] when completed. Delete a line if not applicable. -->
  - [] I have performed a self-review of my code
  - [] I have added docstrings/doxygen in line with the guidance in the developer guide
  - [] I have implemented unit tests that cover any new or modified functionality (if applicable)
  - [] The code builds and runs on my machine
  - [] `documentation/release_XXX.md` has been updated with any functionality change (if applicable)

NikEfth committed:
Initial PyTorch support

NikEfth committed:
A bit of a clean up and example in the stir_math (it will be removed in the future)

KrisThielemans commented on src/cmake/stir_exe_targets.cmake:
```diff
@@ -22,6 +22,10 @@ foreach(source ${${dir_EXE_SOURCES}})
   if(BUILD_EXECUTABLES)
    get_filename_component(executable ${source} NAME_WE)
    add_executable(${executable} ${ALL_HEADERS} ${ALL_INLINES}  ${ALL_TXXS} ${source} $<TARGET_OBJECTS:stir_registries>)
+if (STIR_WITH_TORCH)
+   target_link_libraries(${executable}  ${TORCH_LIBRARIES})
+   target_include_directories(${executable} PUBLIC ${TORCH_INCLUDE_DIRS})
+endif()
```
irrelevant comment for now, but do we really need this for all exes? The mod to src/buildblock/CMakeLists.txt should be enough (due to transitive dependence)

NikEfth replied:
Probably not ... but let's see :)

KrisThielemans commented on src/cmake/stir_test_exe_targets.cmake:
```diff
@@ -59,6 +59,10 @@ macro (create_stir_involved_test source  libraries dependencies)
  if(BUILD_TESTING)
    get_filename_component(executable ${source} NAME_WE)
    add_executable(${executable} ${source} ${dependencies})
+   if (STIR_WITH_TORCH)
+      target_link_libraries(${executable}  ${TORCH_LIBRARIES})
+      target_include_directories(${executable} PUBLIC ${TORCH_INCLUDE_DIRS})
+   endif()
```
as in src/cmake/stir_exe_targets.cmake

NikEfth committed:
PyTorch STIR introduction and the playground utility.

NikEfth committed:
More wrapped functions and some cleanup

NikEfth committed:
Separate TensorWrapper class.
* On going work to make SWIG happy.

NikEfth committed:
From functionality in the TensorWrapper

NikEfth committed:
Comment out most Tensor's functionality until we have a better idea of how to use it

NikEfth committed:
Merge remote-tracking branch 'origin/master' into STIR_pytorch

NikEfth committed:
Voxels- Pixels OnCartesian Grid are working with Tensor.
* Almost ready to ready the data in Tensors.
* Implemented Tensor Iterators, strictly for CPU usage.

KrisThielemans commented on src/include/stir/DiscretisedDensity.inl:
```diff
@@ -39,16 +39,25 @@ DiscretisedDensity<num_dimensions, elemT>::DiscretisedDensity()
                                                               const CartesianCoordinate3D<float>& origin_v)
     : ExamData(exam_info_sptr),
-      Array<num_dimensions, elemT>(range_v),
+#ifdef STIR_WITH_TORCH
+    TensorWrapper<num_dimensions, elemT>(range_v),
+#else
+    Array<num_dimensions, elemT>(range_v),
+#endif
```
all of these ifdefs could be replaced by `base_type(range_v)`

NikEfth committed:
MINI STIR compiles.
A lot of work in the IO and buildblock.
We should be able to read and write all the data types in the IO module,
in Tensors and hopefully have not broken anything for Array in the process.
* My confusion comes because I don't know, now how many of these fucntions are
used by ProjData.

NikEfth commented:
This is where we are now: 

```
Running tests...
Test project /local_mount/space/celer/2/users/pytorch_playground/STIR.b
    Start 1: test_Array
1/2 Test #1: test_Array .......................   Passed    2.58 sec
    Start 2: test_VectorWithOffset
2/2 Test #2: test_VectorWithOffset ............   Passed    0.50 sec

100% tests passed, 0 tests failed out of 2```
- I know, it says absolutely nothing :)

NikEfth committed:
* Reverting reading the Interfile header with vectors of floats.
It is a bigger headache than I thought.

NikEfth committed:
The first of the most interesting commits to STIR_pytorch.
We start the tensor tests.
Access the tensor elements through the .at() method, that ensures
that best speed and efficiency in CPU.
GPU access should allowed only be with linear operators.

NikEfth committed:
More functionalities for TensorWrapper.
* test_Tensor replicates the tests from test_Array and is mostly ready for:
    - testing for simple operations with and without offsets
    - 1D and 2D (somewhat work needed)

NikEfth commented:
Made quite a bit of progress, a few more boring parts left: 
```
Running tests...
Test project /local_mount/space/celer/2/users/pytorch_playground/STIR.b
    Start 1: test_Array
1/3 Test #1: test_Array .......................   Passed    8.19 sec
    Start 2: test_VectorWithOffset
2/3 Test #2: test_VectorWithOffset ............   Passed    0.49 sec
    Start 3: test_Tensor
3/3 Test #3: test_Tensor ......................   Passed    0.50 sec

100% tests passed, 0 tests failed out of 3

Total Test time (real) =   9.18 sec
```

KrisThielemans reviewed (changes requested):
what about creating an alias to minimise `#ifdef STIR_TORCH` more. ATM, there's quite a few non-changes because of this. e.g.

ArrayTypeFwd.h
```
namespace stir
{

template <typename elemT, int num_dimensions>
class Array;
#ifndef STIR_TORCH
template <typename elemT, int num_dimensions>
using ArrayType = Array<elemT, num_dimensions>
#else
template <typename elemT, int num_dimensions>
TensorWrapper;
template <typename elemT, int num_dimensions>
using ArrayType = TensorWrapper<elemT, num_dimensions>
#endif
}
```

and ArrayType.h
```
#include "stir/ArrayTypeFwd.h"
#include "stir/Array.h"
#ifdef STIR_TORCH
#include "stir/TensorWrapper.h"
#endif
```
then use `ArrayType` wherever you want to replace `Array` with `TensorWrapper`

I'm not 100% sure about `using` syntax though.

KrisThielemans commented:
See #1589. Possibly move some stuff from here to there if I didn't cover everything yet. Let's merge #1589 quickly though (before 6.3).

NikEfth commented:
@KrisThielemans 
Hi, I wrote the 2D check next() test for the TensorWrapper. 
Can you please confirm that this is correct (BasicCoordinate always gives me doubts): 

```
std::cerr << "\tTest on next() with regular array\n";
          BasicCoordinate<2, int> index;
          index[1] = test.get_min_index(0);
          index[2] = test.get_min_index(1);
          cerr << ">>>>" << index[1] << "  " << index[2] << std::endl;
          TensorWrapper<2, int>::const_full_iterator iter = test.begin_all_const();
          do
            {
              cerr << ">>" << index[1] << "  " << index[2] << std::endl;
              check(*iter == test.at(index), "test on next(): element out of sequence?");
              ++iter;
              index[2]+=1;
              if (index[2] > test.get_max_index(1))
                {
                  index[2] = test.get_min_index(1);
                  index[1]+=1;
                }
          } while (iter != test.end_all());
          check(iter == test.end_all(), "test on next() : did we cover all elements?");
        }
      }
```

KrisThielemans commented:
Looks ok, but I guess I'd write it with a loop in terms of the index (which is closer to how STIR will be using it)
```c++
auto iter = test.begin_all_const();
for (index[1] = test.get_min_index(0); index[1] <= test.get_max_index(0); ++index[1])
  for (index[2] = test.get_min_index(1); index[2] <= test.get_max_index(1); ++index[2])
    {
       check(*iter == test.at(index), "test on next(): element out of sequence?");
       ++iter;
    }
```
The choice to start from 1 was a very bad one... At some point, we'll have to introduce a new index-type that starts from 0 (could probably just use `std::array`), maybe auto-converting the `BasicCoordinate` to the new type.

NikEfth commented:
> Looks ok, but I guess I'd write it with a loop in terms of the index (which is closer to how STIR will be using it)

I know. I wanted to replicate closely the spelling of that test. it is based on a similar do .. while

NikEfth commented:
@KrisThielemans  in NumericVector operators you have this line
``` 
  this->grow(std::min(this->get_min_index(), v.get_min_index()), std::max(this->get_max_index(), v.get_max_index()));
```
This automatically resizes the vector if it is smaller. 
Do you want this behavior?
Shouldn't the user handle this?

NikEfth commented:
A new PR is coming. closing this
